lib/vagrant-service-manager/services/docker.rb

```diff
@@ -71,6 +69,35 @@ def self.print_env_info(ui, options)
           PluginUtil.print_shell_configure_info(ui, ' docker')
         end
       end
+
+      def self.install_cli(machine, ui, options)
+        Installer.new('docker', machine, ui, options).run
+      end
+
+      def self.binary_path(version)
+        "#{BIN_FOLDER}/docker/#{version}/docker"
+      end
+
+      def self.binary_download_url(machine, version)
+        arch = machine.env.host.capability(:os_arch)
+        "#{BINARY_BASE_URL}/#{os_type}/#{arch}/docker-#{version}"
```

Comment:
But you will end up with switch statement in the service. Exactly what we tried to avoid. In this case  why did you create these executors in the first place. Also a `service` imo does not have to have knowledge over where client libs are coming from. And how do you know that the CDK will have a download URL. The mechanism might be completely different in which case things get even more complicated. The how of getting the cli libs should completely be encapsulated in the executors. Your abstractions are leaking.